### PR TITLE
Update regex for date and datetime validation

### LINKED ISSUE #5230: Increase validations on CSV upload fields
a couple of our uploads failed with no useful error message to the user because they hit some RS validations. see https://skylight-hq.slack.com/archives/C01HFHPNJEL/p1675787314283249?thread_ts=1675112280.908719&cid=C01HFHPNJEL

Looks like pretty basic errors. Two of  the 400's were for incorrect DOB format (received xx/0/xxxx). Another was for an equipment_model_name not matching to value on the LIVD table. Not sure about the 503 yet. Possibly just a transient error  if we haven't seen it recur.

ask:
* update the [date regex](https://github.com/CDCgov/prime-simplereport/blob/ea63620064045386e6ab64c0267fc00bfec85eec/backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java#L39) to not allow invalid dates in the format `xx/0/xxxx`. our date time regex should also be updated as well

## Acceptance criteria
- update the [date regex](https://github.com/CDCgov/prime-simplereport/blob/ea63620064045386e6ab64c0267fc00bfec85eec/backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java#L39) to not allow invalid dates in the format `xx/0/xxxx`. our date time regex should also be updated as well (don't allow zero only for M or D)

Estimate: 1

### PR DESCRIPTION
**Pull request by mpbrown**
# BACKEND PULL REQUEST

## Related Issue

- Resolves #5230 

## Changes Proposed

- Updates regex for date and datetime so that month values are restricted to 1 to 12 and day values are restricted to 1 to 31. Values of 01 to 09 are also still valid for month and day.

## Additional Information

- The specific acceptance criteria for this issue is not allowing zeros only for month or day. This resolves the issue for zeros as well as validating the maximum allowed values for month and day.
- Current allowed date formats are `MM/DD/YYYY` and `M/D/YYYY`
- Current allowed datetime formats are `MM/DD/YYYY HH:mm`, `MM/DD/YYYY H:mm`, `M/D/YYYY HH:mm`, or `M/D/YYYY H:mm`
- The previous date regex allowed for it, but just to confirm are these also acceptable formats? `M/DD/YYYY` or `MM/D/YYYY`
- When validating dates and datetimes, should we also be checking for invalid/non-existent dates such as April 31 or handling leap years for February 29?

## Testing

- Look over and run the new tests
- See if there are any other test cases or edge cases that need to be handled

### CODE REVIEW

File: backend/src/test/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtilsTest.java [hunk 1]
```diff
@@ -135,4 +138,66 @@ void invalidSpecimenSNOMED() {
...
+    validDates.add(new ValueOrError("1/31/2023", "date"));
+    validDates.add(new ValueOrError("12/01/2023", "date"));
+    for (var date : validDates) {
+      assertThat(validateDateFormat(date)).isEmpty();
+    }
+  }
+
+  @Test
+  void invalidDateFormat() {
+    var validDates = new ArrayList<ValueOrError>();
```
Comment (zdeveloper) [inline backend/src/test/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtilsTest.java:158]: I would rename the variable to `invalidDates`

File: backend/src/test/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtilsTest.java [hunk 2]
```diff
@@ -135,4 +138,66 @@ void invalidSpecimenSNOMED() {
...
+    validDateTimes.add(new ValueOrError("1/31/2023 05:50", "datetime"));
+    validDateTimes.add(new ValueOrError("12/01/2023 1:01", "datetime"));
+    for (var datetime : validDateTimes) {
+      assertThat(validateDateTime(datetime)).isEmpty();
+    }
+  }
+
+  @Test
+  void invalidDateTime() {
+    var validDateTimes = new ArrayList<ValueOrError>();
```
Comment (zdeveloper) [inline backend/src/test/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtilsTest.java:189]: same as before rename to `invalidDates`

**Review by zdeveloper — APPROVED** [review]
LGTM! Thank you for the in-depth testing

File: backend/src/test/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtilsTest.java (hunk 1 above)
Comment (mpbrown) [reply backend/src/test/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtilsTest.java:158]: Good catch on the copy and paste!

**Review by DanielSass — APPROVED** [review]

**Review by zdeveloper — APPROVED** [review]
Thank you!

**Comment by mpbrown** [comment]
@zdeveloper @DanielSass outside of the regex validation, should we also handle invalid/non-existent dates such as April 31 or leap years for February 29?

**Comment by DanielSass** [comment]
> @zdeveloper @DanielSass outside of the regex validation, should we also handle invalid/non-existent dates such as April 31 or leap years for February 29?

I'd be comfortable with not 🤷
